Moved common API response code into extension methods.

izevaka opened the pull request:
- Added QueryString extension method that tokenizes URL querystring
  parameters.
- Added TryGetAndConvert method that gets a querystring parameter and tries
  to convert it to a specified type
- Created a whole bunch of extension methods to return success and error
  responses.

izevaka committed:
Moved common API response code into extension methods.

- Added QueryString extension method that tokenizes URL querystring
  parameters.
- Added TryGetAndConvert method that gets a querystring parameter and tries
  to convert it to a specified type
- Created a whole bunch of extension methods to return success and error
  responses.

davidfowl commented on JabbR/api/MessagesController.cs:
```diff
@@ -20,7 +21,7 @@ public MessagesController(IJabbrRepository repository)
 
-        public HttpResponseMessage GetAllMessages(string room, string range, bool download = false)
+        public HttpResponseMessage GetAllMessages(string room, string range)
         {
```
Where's the download parameter?

izevaka replied:
That's now in the common bit here - https://github.com/davidfowl/JabbR/pull/490/files#diff-3.

The thinking is that download parameter can be applied to any api call, so I wanted to move it out of any specific controllers, to keep them cleaner.

davidfowl replied:
Ok this looks ok.

davidfowl commented on JabbR/Infrastructure/UriExtensions.cs:
```diff
@@ -0,0 +1,96 @@
+
+namespace JabbR.Infrastructure
+{
+    public static class UriExtensions
```
Why do you need this class?

izevaka replied:
Because outside of action, parameter/model binding is not available. I would imagine that we'd need this when we need to parse api key (probably in an authorization filter. Also when adding jsonp support. Not sure where jsonp bit would sit, probably needs a customized mediaformatter with a bit of code in delegatingmessagehandler to set content-type.

davidfowl replied:
Can we add it when we need it? I don't like things being there for the sake of it just in case we're going to need it. Bring it back when it's being used.

izevaka replied:
It's already used now by AddDownloadHeader. I think I know what to do to make it a bit cleaner. How about creating a querystring collection class, that way, all the semantics of getting a value and converting it that are very specific to url parameter binding are not available on a generic IDictionary extension.

izevaka committed:
Refactored UriExtensions to be a non-static class.

- Created a QueryString class that wraps up extracting and converting query
  string parameters.

davidfowl commented on JabbR.Tests/QueryStringCollectionTests.cs:
```diff
@@ -0,0 +1,154 @@
+            QueryStringCollection _QueryStringCollection;
+            public TryGetAndConvert()
+            {
+                _QueryStringCollection = new QueryStringCollection(new Uri("http://example.com?intvalue=1&stringvalue=str&boolvalue=true&empty"));
```
I'm not a fan of class level arguments in unit tests, can you move this to each test? It makes it clearer as you don't have to look outside the test to get context.

izevaka committed:
Removed common setup from unit test constructors.

izevaka committed:
Cleaned up variable names for consistency sake.

- Changed _QueryStringCollection local variable to queryStringCollection.
- Changed _QueryStringDictionary member variable to _queryStringDictionary.

davidfowl commented:
Rebased.
